### PR TITLE
8340796: Use a consistent order when loading cxq and EntryList

### PR DESCRIPTION
**Pull request by fbredber**
Made sure we use a consistent order when reading ObjectMonitor EntryList and cxq while exiting the monitor.
That consistent order is EntryList before cxq. Added a comment in the shared code to explaining why it's okay that waiters slip undetected between reading EntryList and cxq.

Tested ok tier1-3.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8340796](https://bugs.openjdk.org/browse/JDK-8340796): Use a consistent order when loading cxq and EntryList (**Enhancement** - P4)

### Reviewers
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**) Review applies to [21da21e5](https://git.openjdk.org/jdk/pull/21584/files/21da21e51c6bf5c9ab211d2c6b6945f350ccbe83)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21584/head:pull/21584` \
`$ git checkout pull/21584`

Update a local copy of the PR: \
`$ git checkout pull/21584` \
`$ git pull https://git.openjdk.org/jdk.git pull/21584/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21584`

View PR using the GUI difftool: \
`$ git pr show -t 21584`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21584.diff">https://git.openjdk.org/jdk/pull/21584.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21584#issuecomment-2423151371)

### CODE REVIEW

**Comment by dean-long** [comment]
We don't have a LoadLoad on aarch64 to enforce the ordering, so isn't it more correct to say that the code handles either ordering correctly, and leave x86 the way it was?

**Review by fisk — APPROVED** [review]
Looks good to me.

File: src/hotspot/share/runtime/objectMonitor.cpp [hunk 1]
```diff
@@ -1133,6 +1133,30 @@ void ObjectMonitor::UnlinkAfterAcquire(JavaThread* current, ObjectWaiter* curren
...
+//      waiting for the monitor, even though thread 2 has waited for the
+//      monitor since before thread 3 released it.
+//
+// However the successor protocol deals with this just fine. Since thread 3
+// managed to enter the monitor, the responsibility of ensuring liveness of
+// the monitor becomes the responsibility of thread 3 which will make
+// thread 2 the successor and unpark thread 2 before exiting the monitor.
+// The race could be removed by reading cxq before EntryList, but it
+// would come with an added cost of needing a loadload fence between the
+// reads.
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/objectMonitor.cpp:1158]: I think you can write this much more succinctly as this isn't really a race per-se, but part of the overall operation of the protocol. Suggestion:
```
// Note that we read the EntryList and then the cxq after dropping the lock, so the values need not form a stable
// snapshot. In particular, after reading the (empty) EntryList, another thread could acquire and release the lock, moving 
// any entries in the cxq to the EntryList, causing the current thread to see an empty cxq and conclude there are no
// waiters. But this is okay as the thread that moved the cxq is responsible for waking the successor.
```

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp [hunk 2]
```diff
@@ -478,8 +478,8 @@ void C2_MacroAssembler::fast_unlock(Register objReg, Register boxReg, Register t
   membar(StoreLoad);
 
   // Check if the entry lists are empty.
```
Comment (dholmes-ora) [inline src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:480]:
```suggestion
  // Check if the entry lists are empty (EntryList first - by convention).
```

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
I'm not really seeing the need for any code change here given that the code works fine whichever order you read them. Having a convention is not unreasonable but unless you document in the comments (in all the platform files) that there is a convention noone will remember it is there.

And given:
```
   if ((intptr_t(_EntryList)|intptr_t(_cxq)) == 0 || _succ != nullptr) {
      return;
    }
```
you don't know for sure which order the compiler will issue the loads anyway.

File: src/hotspot/share/runtime/objectMonitor.cpp (hunk 1 above)
Comment (dholmes-ora) [reply src/hotspot/share/runtime/objectMonitor.cpp:1158]: Update: okay the fields are volatile so maybe the compiler will preserve the written order.

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp (hunk 2 above)
Comment (fbredber) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp]: fixed

File: src/hotspot/share/runtime/objectMonitor.cpp (hunk 1 above)
Comment (fbredber) [reply src/hotspot/share/runtime/objectMonitor.cpp]: I'd decided to go with your suggestion. Thanks!

**Comment by fbredber** [comment]
@dean-long 
> We don't have a LoadLoad on aarch64 to enforce the ordering, so isn't it more correct to say that the code handles either ordering correctly, and leave x86 the way it was?

Both you and David was on to the same thing, so I decided to rewrite as per David's suggestion. However I will change the x86 code so it will follow the same order as the rest, just to avoid any head scratching about why x86 is done differently.

**Comment by dean-long** [comment]
If reading cxq first avoids the benign "race", then why not make that the preferred ordering convention?

It sounds like we are not too concerned about avoiding this particular protocol state (if we aren't calling it a race), but if we were, there are platform-specific things we could try.  For example, some platforms should be able to read both fields at the same time if the fields are adjacent in memory and suitably aligned.

**Review by coleenp — APPROVED** [review]
Even though the order isn't strictly important, having the convention to read one before the other avoids the situation where we wonder whether someone's already looked at this to work out whether it's safe to do.  Note this came up in a code review and rather than wondering again whether this is good at some later date, I like that it's in the same order with the comment that explains that load ordering is unnecessary.

**Review by dholmes-ora — APPROVED** [review]
LGTM!

**Pull request closed without merging**